xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -133,6 +145,61 @@ bool CalculatesExpectation(const HloInstruction* instr) {
          ((actual_r_nelems + r_nelems) * numerical_epsilon);
 }
 
+// Recursively traverses the graph across converts, bitcasts, reshapes and
+// transposes, starting from instr, and returns the operand of custom_call with
+// index operand_idx. Returns nullptr if the prescribed operand of custom_call
+// is not found.
+HloInstruction* FindOperandRecursive(HloInstruction* instr,
+                                     HloInstruction* custom_call,
+                                     absl::flat_hash_set<int>& visited_instrs,
+                                     int operand_idx) {
+  visited_instrs.emplace(instr->unique_id());
+  const absl::flat_hash_set<HloOpcode> supported_ops = {
+      HloOpcode::kConvert, HloOpcode::kBitcast, HloOpcode::kReshape,
+      HloOpcode::kTranspose};
+  // Look for the Custom Call among the users of instr.
+  for (HloInstruction* user : instr->users()) {
+    if (user->unique_id() == custom_call->unique_id() &&
+        user->operand_index(instr) == operand_idx) {
+      return instr;
+    }
+    if (supported_ops.contains(user->opcode()) &&
+        !visited_instrs.contains(user->unique_id())) {
+      return FindOperandRecursive(user, custom_call, visited_instrs,
+                                  operand_idx);
+    }
+  }
+  // Ascend the graph if the Custom Call is not found and instr is a
+  // conversion, reshape or transpose.
+  if (supported_ops.contains(instr->opcode())) {
+    return FindOperandRecursive(instr->mutable_operand(0), custom_call,
+                                visited_instrs, operand_idx);
+  }
+  return nullptr;
+}
+
+// Recursively traverses the graph downward across converts, bitcasts, reshapes
+// and transposes, starting from instr, and returns the first addition-reduction
+// identified. Returns nullptr if no addition-reduction is found.
+HloInstruction* FindAddReduceRecursive(
+    HloInstruction* instr, absl::flat_hash_set<int>& visited_instrs) {
+  visited_instrs.emplace(instr->unique_id());
+  const absl::flat_hash_set<HloOpcode> supported_ops = {
+      HloOpcode::kConvert, HloOpcode::kBitcast, HloOpcode::kReshape,
+      HloOpcode::kTranspose};
+  // Look for the reduction among the users of instr.
+  for (HloInstruction* user : instr->users()) {
+    if (AppliesAddReduce(user)) {
```

Comment:
As in other places, I think you need to check axis, and also any transposes you find can change the effective axis.